File: LeetCode/Find Kth Bit in Nth Binary String.swift
/*
  Find Kth Bit in Nth Binary String
  URL: https://leetcode.com/problems/find-kth-bit-in-nth-binary-string
  Difficulty: Medium
  Topic : String, Recursion, Simulation
*/

import Foundation
class Solution {
    func findKthBit(_ n: Int, _ k: Int) -> Character {
        if n == 1 { return "0" }
        var i = 1
        var count = 0
        var kCopy = k
        while Int(pow(2.0,Double(i))) <= k {
            i += 1
        }
        while true {
            let center = Int(pow(2.0,Double(i-1)))
            if center == kCopy {
                let base = Character(i == 1 ? "0" : "1")
                let invert = Character(i == 1 ? "1" : "0")
                return count % 2 == 0 ? base : invert
            }
            if center < kCopy {
                kCopy = Int(pow(2.0,Double(i))) - kCopy
                count += 1
            }                
            i -= 1
        }
        return "1"
    }
}

File: LeetCode/Python/Find Kth Bit in Nth Binary String.py
"""
  Find Kth Bit in Nth Binary String
  URL: https://leetcode.com/problems/find-kth-bit-in-nth-binary-string
  Difficulty: Medium
  Topic : String, Recursion, Simulation
"""

class Solution:
    def findKthBit(self, n: int, k: int) -> str:
        if n == 1: return "0"
        i = 1

        while 2 ** i <= k:
            i += 1
        
        count = 0
        while True:
            if (2 ** (i-1)) == k:
                base = "0" if i == 1 else "1"
                invert = "1" if i == 1 else "0"
                return (base if count % 2 == 0 else invert)
            if (2 ** (i - 1)) < k:
                k = (2 ** i) - k
                count += 1
            i -= 1
        
        return "0"
        
